// @blitzjs/cli/lib/src/commands/prisma.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrismaCommand = exports.runPrismaExitOnError = exports.runPrisma = void 0;
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
// @blitzjs/server imports react, so we must import the @blitzjs/server version of the
// local app instead of the global.
// import-cwd is required so this works correctly during new app generation
const getPrismaBin = () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    let bin;
    try {
        bin = require("import-cwd")("@blitzjs/server").resolveBinAsync("prisma", "prisma");
    }
    catch (_a) {
        // legacy compatability
        bin = require("import-cwd")("@blitzjs/server").resolveBinAsync("@prisma/cli", "prisma");
    }
    return bin;
});
let prismaBin;
const runPrisma = (args, silent = false) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    if (!prismaBin) {
        try {
            prismaBin = yield getPrismaBin();
        }
        catch (err) {
            throw err;
        }
    }
    const cp = require("cross-spawn").spawn(prismaBin, args, {
        stdio: silent ? "ignore" : "inherit",
        env: process.env,
    });
    const code = yield require("p-event")(cp, "exit", { rejectionEvents: [] });
    return code === 0;
});
exports.runPrisma = runPrisma;
const runPrismaExitOnError = (...args) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const success = yield exports.runPrisma(...args);
    if (!success) {
        process.exit(1);
    }
});
exports.runPrismaExitOnError = runPrismaExitOnError;
class PrismaCommand extends command_1.Command {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { argv } = this.parse(PrismaCommand);
            yield exports.runPrismaExitOnError(argv);
        });
    }
}
exports.PrismaCommand = PrismaCommand;
PrismaCommand.description = "Loads env variables then proxies all args to Prisma CLI";
PrismaCommand.aliases = ["p"];
PrismaCommand.strict = false;
