// @blitzjs/cli/lib/src/commands/start.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Start = void 0;
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
class Start extends command_1.Command {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { flags } = this.parse(Start);
            const config = {
                rootFolder: process.cwd(),
                port: flags.port,
                hostname: flags.hostname,
                inspect: flags.inspect,
                clean: true,
                env: "prod",
            };
            try {
                const prod = (yield Promise.resolve().then(() => tslib_1.__importStar(require("@blitzjs/server")))).prod;
                yield prod(config);
            }
            catch (err) {
                console.error(err);
                process.exit(1); // clean up?
            }
        });
    }
}
exports.Start = Start;
Start.description = "Start the production server";
Start.aliases = ["s"];
Start.flags = {
    help: command_1.flags.help({ char: "h" }),
    port: command_1.flags.integer({
        char: "p",
        description: "Set port number",
    }),
    hostname: command_1.flags.string({
        char: "H",
        description: "Set server hostname",
    }),
    inspect: command_1.flags.boolean({
        description: "Enable the Node.js inspector",
    }),
};
