// @blitzjs/core/dist/blitz-data-f63b7a88.cjs.dev.js
'use strict';

var config = require('@blitzjs/config');
var htmlescape = require('htmlescape');
var React = require('react');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var htmlescape__default = /*#__PURE__*/_interopDefault(htmlescape);
var React__default = /*#__PURE__*/_interopDefault(React);

var isServer = typeof window === "undefined";
var isClient = typeof window !== "undefined";
function clientDebug() {
  if (typeof window !== "undefined" && window["DEBUG_BLITZ"]) {
    var _console;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    (_console = console).log.apply(_console, ["[BLITZ]"].concat(args));
  }
}

function _getBlitzRuntimeData() {
  var _config$middleware, _middleware$config, _config$experimental;

  var config$1 = config.getConfig();
  var middleware = (_config$middleware = config$1.middleware) == null ? void 0 : _config$middleware.filter(function (middleware) {
    return middleware.name === "blitzSessionMiddleware";
  })[0];
  var cookiePrefix = middleware == null ? void 0 : (_middleware$config = middleware.config) == null ? void 0 : _middleware$config.cookiePrefix;
  return {
    sessionCookiePrefix: cookiePrefix || "blitz",
    suspenseEnabled: ((_config$experimental = config$1.experimental) == null ? void 0 : _config$experimental.reactRoot) !== false
  };
}
function getBlitzRuntimeData() {
  if (isClient && !process.env.JEST_WORKER_ID) {
    return window.__BLITZ_DATA__;
  } else {
    return _getBlitzRuntimeData();
  }
} // Automatically deserialize __BLITZ_DATA__ in a browser environment

if (isClient && !process.env.JEST_WORKER_ID) {
  deserializeAndSetBlitzDataOnWindow();
}

function deserializeAndSetBlitzDataOnWindow() {
  try {
    var data = JSON.parse(document.getElementById("__BLITZ_DATA__").textContent);
    window.__BLITZ_DATA__ = data;
  } catch (e) {
    console.error("Error deserializing __BLITZ__DATA__. Make sure you have a custom _document.js/tsx page that uses <BlitzScript/>", e);
  }
}
function BlitzData() {
  return /*#__PURE__*/React__default['default'].createElement("script", {
    id: "__BLITZ_DATA__",
    type: "application/json",
    dangerouslySetInnerHTML: {
      __html: htmlescape__default['default'](_getBlitzRuntimeData())
    }
  });
}

exports.BlitzData = BlitzData;
exports.clientDebug = clientDebug;
exports.getBlitzRuntimeData = getBlitzRuntimeData;
exports.isClient = isClient;
exports.isServer = isServer;
